src/DynamicNLPModels.jl

```diff
@@ -4,42 +4,597 @@ import NLPModels
+
+The following keyward arguments are also accepted
+- `Qf`: objective function matrix for system state at time N; defaults to Q
+- `sl = fill(-Inf, ns)`: vector of lower bounds on state variables
+- `su = fill(Inf, ns)` : vector of upper bounds on state variables
+- `ul = fill(-Inf, nu)`: vector of lower bounds on input variables
+- `uu = fill(Inf, nu)` : vector of upper bounds on input variables
+"""
+function LQDynamicData(
+    s0::S,
+    A::M,
+    B::M,
+    Q::M,
+    R::M,
+    N::Int;
+
+    Qf::M = Q, 
+    sl::S = (similar(s0) .= -Inf),
+    su::S = (similar(s0) .=  Inf),
+    ul::S = (similar(s0,size(R,1)) .= -Inf),
+    uu::S = (similar(s0,size(R,1)) .=  Inf)
+    ) where {T,S <: AbstractVector{T},M <: AbstractMatrix{T}}
+
+    if size(Q,1) != size(Q,2) 
+        error("Q matrix is not square")
+    end
+    if size(R,1) != size(R,1)
+        error("R matrix is not square")
+    end
+    if size(A,2) != length(s0)
+        error("Number of columns of A are not equal to the number of states")
+    end
+    if size(B,2) != size(R,1)
+        error("Number of columns of B are not equal to the number of inputs")
+    end
+    if length(s0) != size(Q,1)
+        error("size of Q is not consistent with length of x0")
+    end
+
+    if !(sl  <= su)
+        error("lower bound(s) on x is > upper bound(s)")
+    end
+    if !(ul <= uu)
+        error("lower bound(s) on u is > upper bound(s)")
+    end
+    if !(s0 >= sl) || !(s0 <= su)
+        error("x0 is not within the given upper and lower bounds")
+    end
+
+    ns= size(Q,1)
+    nu= size(R,1)
+
+
+    LQDynamicData{T,S,M}(
+        s0, A, B, Q, R, N,
+
+        Qf, ns, nu,
+
+        sl, su, ul, uu 
+    )
+end
+"""
+    LQDynamicData(ns::Int, nu::Int, N::Int) -> LQDynamicData{T, S, M}
+
+A constructor for building an object of type `LQDynamicData` for the optimization problem 
+
+    minimize    1/2 sum(si^T Q si for i in 1:(N-1)) + 1/2 sum(ui^T R ui for i in 1:(N-1)) + 1/2 sN^T Qf s_N
+    subject to  s{i+1} = A s{i} + B u{i}  for i in 1:(N-1)
+                sl <= s <= su
+                ul <= u <= uu
+                s{1} = s0
+
+---
+
+Initializes empty `Q`, `R`, `A`, and `B` (sparse) arrays and empty `s0`, `sl`, `su`, `ul`, and `uu` vectors
+
+Inputs:
+- `ns`: number of states
+- `nu`: number of inputs
+- `N` : number of time points
+"""
+function LQDynamicData(ns::Int, 
+    nu::Int, 
+    N::Int
+    )
     
-    A::MT
-    B::MT
-    Q::MT
-    R::MT
-    Qf::MT
+    Q  = SparseArrays.sparse([],[],Float64[], ns, ns)
+    R  = SparseArrays.sparse([],[],Float64[], nu, nu)
+    A  = SparseArrays.sparse([],[],Float64[], ns, ns)
+    B  = SparseArrays.sparse([],[],Float64[], ns, nu)
+    Qf = SparseArrays.sparse([],[],Float64[], ns, ns)
+
+    s0 = zeros(ns)
+    sl = fill(-Inf, ns)
+    su = fill(Inf, ns)
+    ul = fill(-Inf, nu)
+    uu = fill(Inf, nu)
+
+    LQDynamicData(s0, A, B, Q, R, N; Qf = Qf, sl = sl, su = su, ul = ul, uu = uu)
+
+end
+
+
+
```

Comment:
for these functions, see if you can use @eval to make it more concise. See:
https://github.com/JuliaSmoothOptimizers/NLPModels.jl/blob/main/src/nlp/tools.jl#L5-L17

Reply:
this will make your job of maintaining the code much easier ;)